File: lib/kota.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import 'supabase_client.dart';

/// Web `app/lib/kota.ts` ile aynı: bir hesap son 30 günde en fazla 3 eşya alabilir.
const aylikAlmaHakki = 3;

class KotaDurumu {
  final int alinan;
  final int kalan;
  final DateTime? yenilenmeTarihi;

  const KotaDurumu({
    required this.alinan,
    required this.kalan,
    this.yenilenmeTarihi,
  });
}

/// Bir hesabın son 30 gündeki eşya alma kotası durumu.
Future<KotaDurumu> kotaDurumu(String uid) async {
  int alinan = 0;
  try {
    final sayi = await supabase.rpc('alinan_esya_sayisi', params: {'kisi': uid});
    if (sayi is int) alinan = sayi;
  } on PostgrestException {
    // kota-limiti.sql henüz çalıştırılmadıysa RPC yok — kotayı boş say
  }

  final kalan = (aylikAlmaHakki - alinan) < 0 ? 0 : aylikAlmaHakki - alinan;

  DateTime? yenilenmeTarihi;
  if (kalan == 0) {
    try {
      final tarih =
          await supabase.rpc('kota_yenilenme_tarihi', params: {'kisi': uid});
      if (tarih is String) yenilenmeTarihi = DateTime.tryParse(tarih);
    } on PostgrestException {
      // yoksa geç
    }
  }

  return KotaDurumu(alinan: alinan, kalan: kalan, yenilenmeTarihi: yenilenmeTarihi);
}

/// Belirli bir kişinin son 30 günde aldığı eşya sayısı (bağış alıcı listesi için).
Future<int> alinanEsyaSayisi(String uid) async {
  try {
    final data = await supabase.rpc('alinan_esya_sayisi', params: {'kisi': uid});
    if (data is int) return data;
  } on PostgrestException {
    // yoksa 0
  }
  return 0;
}

const _aylar = [
  'Ocak', 'Şubat', 'Mart', 'Nisan', 'Mayıs', 'Haziran',
  'Temmuz', 'Ağustos', 'Eylül', 'Ekim', 'Kasım', 'Aralık',
];

String tarihMetni(DateTime? tarih) {
  if (tarih == null) return '';
  final t = tarih.toLocal();
  return '${t.day} ${_aylar[t.month - 1]} ${t.year}';
}

File: src/lib/kota.ts
import { supabase } from './supabase';

/** Varsayılan/yedek değer. Asıl kaynak: DB `uygulama_ayarlari` → `aylik_alma_hakki`. */
export const AYLIK_ALMA_HAKKI = 3;

export type KotaDurumu = {
  alinan: number;
  kalan: number;
  limit: number;
  yenilenmeTarihi: Date | null;
};

/** Aylık alma hakkı — `uygulama_ayarlari()` RPC'den (fallback: AYLIK_ALMA_HAKKI). */
export async function aylikAlmaHakki(): Promise<number> {
  try {
    const { data } = await supabase.rpc('uygulama_ayarlari');
    const v = (data as { aylik_alma_hakki?: unknown } | null)?.aylik_alma_hakki;
    const n = typeof v === 'number' ? v : Number(v);
    return Number.isFinite(n) && n > 0 ? n : AYLIK_ALMA_HAKKI;
  } catch {
    return AYLIK_ALMA_HAKKI;
  }
}

/** Bir hesabın son 30 gündeki eşya alma kotası durumu. */
export async function kotaDurumu(uid: string): Promise<KotaDurumu> {
  let alinan = 0;
  {
    const { data, error } = await supabase.rpc('alinan_esya_sayisi', { kisi: uid });
    if (!error && typeof data === 'number') alinan = data;
  }

  const limit = await aylikAlmaHakki();
  const kalan = limit - alinan < 0 ? 0 : limit - alinan;

  let yenilenmeTarihi: Date | null = null;
  if (kalan === 0) {
    const { data, error } = await supabase.rpc('kota_yenilenme_tarihi', { kisi: uid });
    if (!error && typeof data === 'string') {
      const t = new Date(data);
      if (!Number.isNaN(t.getTime())) yenilenmeTarihi = t;
    }
  }

  return { alinan, kalan, limit, yenilenmeTarihi };
}

/** Belirli bir kişinin son 30 günde aldığı eşya sayısı (bağış alıcı listesi için). */
export async function alinanEsyaSayisi(uid: string): Promise<number> {
  const { data, error } = await supabase.rpc('alinan_esya_sayisi', { kisi: uid });
  if (!error && typeof data === 'number') return data;
  return 0;
}

const aylar = [
  'Ocak', 'Şubat', 'Mart', 'Nisan', 'Mayıs', 'Haziran',
  'Temmuz', 'Ağustos', 'Eylül', 'Ekim', 'Kasım', 'Aralık',
];

export function tarihMetni(tarih: Date | null | undefined): string {
  if (!tarih) return '';
  return `${tarih.getDate()} ${aylar[tarih.getMonth()]} ${tarih.getFullYear()}`;
}
